src/services/createRouterHooks.ts

```diff
@@ -1,53 +1,203 @@
+    } catch (error) {
+      if (error instanceof CallbackContextPushError) {
+        return error.response
+      }
+
+      if (error instanceof CallbackContextRejectionError) {
+        return error.response
+      }
+
+      if (error instanceof CallbackContextAbortError) {
+        return error.response
+      }
+
+      throw error
+    }
+
+    return {
+      status: 'SUCCESS',
+    }
+  }
+
+  async function runAfterRouteHooks({ to, from }: AfterHookContext): Promise<AfterRouteHookResponse> {
+    const { global, component } = store
+    const route = getAfterRouteHooksFromRoutes(to, from)
+
+    const allHooks: AfterRouteHook[] = [
+      ...component.onAfterRouteLeave,
+      ...route.onAfterRouteLeave,
+      ...global.onAfterRouteLeave,
+      ...component.onAfterRouteUpdate,
+      ...route.onAfterRouteUpdate,
+      ...global.onAfterRouteUpdate,
+      ...component.onAfterRouteEnter,
+      ...route.onAfterRouteEnter,
+      ...global.onAfterRouteEnter,
+    ]
+
+    try {
+      const results = allHooks.map((callback) => callback(to, {
+        from,
+        reject,
+        push,
+        replace,
+      }))
+
+      await Promise.all(results)
+    } catch (error) {
+      if (error instanceof CallbackContextPushError) {
+        return error.response
+      }
+
+      if (error instanceof CallbackContextRejectionError) {
+        return error.response
+      }
+
+      throw error
+    }
+
+    return {
+      status: 'SUCCESS',
+    }
+  }
+
+  const addBeforeRouteHook: RegisterBeforeRouteHook = ({ lifecycle, timing, depth, hook }) => {
+    const condition = getRouteHookCondition(lifecycle)
+    const hooks = store[timing][lifecycle]
+
+    const wrapped: BeforeRouteHook = (to, context) => {
+      if (!condition(to, context.from, depth)) {
+        return
+      }
+
+      return hook(to, context)
+    }
+
+    hooks.add(wrapped)
+
+    return () => hooks.delete(wrapped)
+  }
+
+  const addAfterRouteHook: RegisterAfterRouteHook = ({ lifecycle, timing, depth, hook }) => {
+    const condition = getRouteHookCondition(lifecycle)
+    const hooks = store[timing][lifecycle]
+
+    const wrapped: AfterRouteHook = (to, context) => {
+      if (!condition(to, context.from, depth)) {
+        return
+      }
+
+      return hook(to, context)
+    }
+
+    hooks.add(wrapped)
+
+    return () => hooks.delete(wrapped)
+  }
+
+  const addRouteHooks: AddRouteHooks = (hooks) => {
+    hooks.onBeforeRouteEnter.forEach((hook) => onBeforeRouteEnter(hook))
+    hooks.onBeforeRouteUpdate.forEach((hook) => onBeforeRouteUpdate(hook))
+    hooks.onBeforeRouteLeave.forEach((hook) => onBeforeRouteLeave(hook))
+    hooks.onAfterRouteEnter.forEach((hook) => onAfterRouteEnter(hook))
+    hooks.onAfterRouteUpdate.forEach((hook) => onAfterRouteUpdate(hook))
+    hooks.onAfterRouteLeave.forEach((hook) => onAfterRouteLeave(hook))
   }
 
   return {
+    runBeforeRouteHooks,
+    runAfterRouteHooks,
+    addBeforeRouteHook,
+    addAfterRouteHook,
```

Comment:
do these need to be part of the API? Seems more like an internal utility to make the public methods DRYer

Reply:
The component level hooks use these methods. We don't expose them on the router itself.